language/move-model/src/number_operation.rs

```diff
@@ -0,0 +1,183 @@
+// Copyright (c) The Diem Core Contributors
+// Copyright (c) The Move Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+//! This file defines types, data structures and corresponding functions to
+//! mark the operation (arithmetic or bitwise) that a variable or a field involves,
+//! which will be used later when the correct number type (int or bv<N>) in the boogie program
+
+use crate::{
+    ast::{PropertyValue, Value},
+    model::{FieldId, FunId, FunctionEnv, ModuleId, StructEnv, StructId},
+    pragmas::{BV_PARAM_PROP, BV_RET_PROP},
+    ty::Type,
+};
+use itertools::Itertools;
+use std::{collections::BTreeMap, str};
+
+static PARSING_ERROR: &str = "error happened when parsing the bv pragma";
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Default, Hash)]
+pub enum NumOperation {
```

Comment:
We should have probably discussed this better before, but I would not have expected this information to end up in the move-model. The model is not specific for the Move Prover. Adding this here is not coherent, since this is rather specific for the prover.

Instead, I would have thought this to become one of the multiple attribute extension points. It could be environment extension, or it could be an annotation.

Reply:
I moved `GlobalNumberOperationState` from `GlobalEnv` to its environment extension. However, in the current implementation, it is hard to move the whole file out of move-model because `ExpInfo` needs to carry `NumOperation`.